feat(posit): enable true constexpr construction from numeric literals

## Problem

The `posit<nbits, es, bt>` template is **not actually constexpr-constructible** from numeric literals, despite the constructors being decorated with `constexpr` / `CONSTEXPRESSION`. This blocks creation of compile-time posit constants and, by extension, blocks the compiler from constant-folding any expression that depends on a posit literal.

## Reproducer

```cpp
#include <universal/number/posit/posit.hpp>
using namespace sw::universal;

constexpr posit<32, 2> p_int(42);     // FAILS
constexpr posit<32, 2> p_float(3.14f); // FAILS
constexpr posit<32, 2> p_double(3.14); // FAILS
```

### gcc-13 (-std=c++20):

```
posit_impl.hpp:1240: error: call to non-'constexpr' function 'sw::universal::convert_(...)'
posit_impl.hpp:1214: error: call to non-'constexpr' function 'float std::frexp(float, int*)'
```

### clang-18 (-std=c++20):

```
posit_impl.hpp:1214: note: non-constexpr function 'frexp' cannot be used in a constant expression
```

## Root causes (refined after deep code analysis)

The blockers form **three layers**, and the deepest layer reaches into a foundational building block shared by every Universal type.

### Layer A - top-level conversion entry points in `posit_impl.hpp`

1. `convert_ieee754()` (line 1206) calls `std::frexp` — not `constexpr` until C++26 (P1383).
2. `convert_ieee754()` (line 1210) calls `std::isnan` / `std::isinf` — not `constexpr` either. (The original issue text mentioned only `frexp`; both must be replaced for the float/double path.)
3. All integer constructors route through `convert_ieee754(static_cast<double>(rhs))` (lines 581-590). So even `posit<32,2>(42)` hits the `frexp` blocker.

### Layer B - the free function `convert_<>()`

The free function `convert_<>()` at `posit_impl.hpp:324` is not marked `constexpr`. Even with Layer A fixed, IEEE-754 conversion still bottoms out here.

### Layer C - blockbinary arithmetic primitives (the real gating item)

Reading `convert_<>()` carefully, its call graph reaches several `blockbinary` operators that are **not** currently `constexpr`. These are used by every number type in the library:

| Symbol | File | Notes |
|--------|------|-------|
| `blockbinary::operator+=` | `blockbinary.hpp` | Multi-limb add with carry; the deepest blocker |
| `blockbinary::operator++` (prefix) | `blockbinary.hpp:210` | Calls `operator+=` |
| `blockbinary::operator<<=` | `blockbinary.hpp:484` | Multi-limb shift |
| `blockbinary::operator\|=` | `blockbinary.hpp` | Bitwise OR over limbs |
| free `truncate(...)` | `blockbinary.hpp` | Bit-loop |
| free `twosComplement(...)` | `blockbinary.hpp:60` | Calls `flip()` (constexpr) + `operator+=` (not) |
| `check_inward_projection_range` | `posit_impl.hpp:93` | Pure integer; trivial constexpr promotion |

What is **already constexpr** and can be reused as-is:

- `calculate_unconstrained_k` (`posit_scale_helpers.hpp:49`)
- `find_msb` (`find_msb.hpp:21`)
- `extractFields` (`extract_fields.hpp:17,26`) — already `BIT_CAST_CONSTEXPR`, the same idiom `cfloat` uses
- `posit::minpos/maxpos/minneg/maxneg` (`posit_impl.hpp:951-974`)
- `posit::setbits(uint64_t)` (line 988), `setzero`, `clear`, `flip`
- `blockbinary::setbit`, `test`, `anyAfter`, `flip`, member `twosComplement`
- `blocksignificand::setbit`, `test`, `anyAfter`

What does work today end-to-end:

- `constexpr posit<32, 2> p(SpecificValue::zero);` - yes
- `constexpr posit<32, 2> p_copy(other_constexpr_posit);` - yes (defaulted copy ctor)
- Default construction `constexpr posit<32, 2> p;` - no, `_block` is intentionally uninitialized to keep posit a trivial type (required by `ReportTrivialityOfType<T>`).

## Proof the pattern is achievable: `cfloat`

`cfloat` already supports true constexpr construction from numeric literals on both gcc and clang:

```cpp
constexpr cfloat<32, 8, uint32_t, true, true, false> p(3.14);  // OK
```

Two design decisions in `cfloat_impl.hpp` make this work:

- Integer assignment routes through `convert_signed_integer` / `convert_unsigned_integer` (line 591-601), which are pure bitwise/integer operations marked `constexpr` (line 2430, 2490).
- Float assignment uses `BIT_CAST_CONSTEXPR convert_ieee754` that bit-casts the IEEE-754 source via `__builtin_bit_cast` (constexpr in gcc/clang/MSVC) and operates on raw bit fields, never touching `std::frexp` / `std::ldexp`.

Critically, cfloat sidesteps Layer C entirely because its encoding is a flat sign+exponent+mantissa bit layout that fits in `uint64_t` for `nbits <= 64`. Posit's run-length regime encoding is more involved, but the same uint64_t-only approach works for the common configurations.

## Why this matters

1. **Performance.** Today every `posit p = 3.14;` runs the full IEEE-754 -> posit conversion pipeline at runtime: scale extraction via `frexp`, fraction-bit walk, `convert_` rounding/regime construction. For inner-loop literals this is a measurable cost. With true `constexpr`, the encoding is a compile-time bit pattern.
2. **Constant folding.** Expressions like `Real two_pi = 2 * pi;` cannot be folded when `Real = posit<32,2>` and `pi` is a runtime-constructed posit. With constexpr literals the compiler can fold these into a single bit pattern.
3. **Static-storage initialization.** `static const posit<32,2> kPi = ...;` currently requires dynamic initialization (runs the conversion at startup). True constexpr makes it `constinit`/`constexpr`-eligible, eliminating static-init order issues.
4. **Template metaprogramming.** Lookup tables, compile-time numerical analysis (e.g., the Teju Jagua-style decimal tables), and `if constexpr` branching on posit values become possible.
5. **Consistency.** `cfloat`, `bfloat16`, and several other Universal types already satisfy this contract. Posit is the most-used type in the library and it is the outlier.
6. **Plug-in replacement contract.** A core Universal value proposition is that custom types should be drop-in replacements for native `float`/`double`. Native floats are constexpr; posit is not. Code that uses `constexpr` literals with `Real = float` cannot be retemplated to `Real = posit<32,2>`.

## Implementation plan (revised: split into 3 sequential PRs)

The original 5-phase plan bundled work that cannot safely land in a single PR. The revised split scopes each PR by **risk surface**.

### PR 1 - Phase 1: integer literal constexpr (this issue)

Self-contained, low risk, no blockbinary changes. Achievable today.

- Add `convert_signed_integer` / `convert_unsigned_integer` to `posit_impl.hpp` modeled on cfloat.
- For `nbits <= 64`, compute the posit encoding using pure `uint64_t` shift/mask, calling `setbits(uint64_t)` directly. Bypass `convert_<>()` entirely (integer input has no fraction, so the rounding/sticky machinery is a no-op).
- For `nbits > 64`, fall back to existing path (no change for that uncommon configuration).
- Promote `check_inward_projection_range` to `constexpr` (trivial 5-line fix; pure integer arithmetic).
- Reroute `operator=(signed char/short/int/long/long long)` and the unsigned counterparts through the new functions.
- Add `static_assert` tests in `static/tapered/posit/api/api.cpp` that lock in the contract and verify bit-pattern equivalence with the runtime path.
- **Outcome**: `constexpr posit<N,es> p(42);` works on gcc and clang for all common configurations. Float/double remain runtime-only (no regression for the existing path).

### PR 2 - blockbinary constexpr promotion (separate issue, must land before PR 3)

This is the gating item for Phases 2-3 and touches a foundational building block used by every Universal type. It deserves its own focused PR, with the full regression suite as a smoke test.

- Mark `blockbinary::operator+=`, `operator<<=`, `operator\|=`, `operator++` as `constexpr`.
- Mark free `truncate(...)` and free `twosComplement(...)` as `constexpr`.
- Mark `convert_<>()` as `constexpr` once blockbinary is clean.
- Run full posit / cfloat / lns / fixpnt / takum / integer regression suites on gcc + clang to catch any silent semantic drift.
- MinGW caveat: validate against the `-fno-ipa-icf` toolchain (the documented MinGW ICF interaction with template instantiation order is not affected by `constexpr` per se, but a multi-instantiation test file should exercise it).

### PR 3 - Phase 2: IEEE-754 literal constexpr (separate issue, depends on PR 2)

- Replace `std::frexp` in `convert_ieee754` with `extractFields` (already `BIT_CAST_CONSTEXPR`).
- Replace `std::isnan` / `std::isinf` at line 1210 with raw-bit checks (`rawExponent == ieee754_parameter<Real>::eallset`, `rawFraction == 0` for inf, etc.) - same idiom cfloat uses.
- Mark `convert_ieee754` as `BIT_CAST_CONSTEXPR`.
- Add `static_assert` tests for float/double construction.
- **`long double` caveat**: on Linux x86-64 with 80-bit extended precision, `extract_fields.hpp:66` falls back to a non-constexpr decoder. The `long double` operator will only be `BIT_CAST_CONSTEXPR` on platforms where `long double == double` (MSVC, MinGW). Document this and accept it.

## Acceptance criteria

PR 1 (this issue):
- [x] `constexpr posit<32, 2> p(42);` compiles and produces the same bits as the runtime version
- [x] `constexpr posit<16, 1>`, `constexpr posit<8, 0>`, `constexpr posit<64, 3>` all work for int sources
- [x] All existing posit conversion / arithmetic tests continue to pass on gcc and clang
- [x] `static_assert` tests in `static/tapered/posit/api/api.cpp` lock in the integer constexpr contract

PR 2 + PR 3 (tracked separately):
- [ ] `constexpr posit<32, 2> p(3.14);` compiles and produces the same bits as the runtime version
- [ ] `convert_<>()` is constexpr; blockbinary arithmetic primitives are constexpr

## Related

- Issue #423 (constexpr `log2` for takum) - same theme but different blocker; takum needs a constexpr transcendental, posit only needs to remove `frexp` and propagate `constexpr` through the conversion pipeline.
- Follow-up issues for PR 2 and PR 3 will be filed once PR 1 lands.

--- a/include/sw/universal/number/posit/posit_impl.hpp
+++ b/include/sw/universal/number/posit/posit_impl.hpp
@@ -90,7 +90,7 @@ constexpr unsigned ES_IS_5 = 5;
 // and thus is too coarse to make this decision.
 // Using the scale directly is the simplest expression of the inward projection test.
 template<unsigned nbits, unsigned es, typename bt>
-bool check_inward_projection_range(int scale) {
+constexpr bool check_inward_projection_range(int scale) {
 	// calculate the min/max k factor for this posit config
 	int posit_size = nbits;
 	int k = scale < 0 ? -(posit_size - 2) : (posit_size - 2);
@@ -574,20 +574,19 @@ public:
 	CONSTEXPRESSION posit(double initial_value)       noexcept : _block{ 0 } { *this = initial_value; }
 
 	// assignment operators for native types
-	// Route integer assignments through convert_ieee754 via double cast.
-	// The blocktriple path has a known issue where round() shifts the hidden
-	// bit below the radix position, causing convert(blocktriple, posit) to
-	// misread it as a fraction bit, producing off-by-one rounding errors.
-	constexpr posit& operator=(signed char rhs)        noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(short rhs)              noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(int rhs)                noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(long rhs)               noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(long long rhs)          noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(char rhs)               noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(unsigned short rhs)     noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(unsigned int rhs)       noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(unsigned long rhs)      noexcept { return convert_ieee754(static_cast<double>(rhs)); }
-	constexpr posit& operator=(unsigned long long rhs) noexcept { return convert_ieee754(static_cast<double>(rhs)); }
+	// Integer assignments route through convert_signed_integer / convert_unsigned_integer,
+	// which are constexpr for nbits <= 64. The previous double-cast route through
+	// convert_ieee754 was not actually constexpr because std::frexp is not constexpr.
+	constexpr posit& operator=(signed char rhs)        noexcept { return convert_signed_integer(rhs); }
+	constexpr posit& operator=(short rhs)              noexcept { return convert_signed_integer(rhs); }
+	constexpr posit& operator=(int rhs)                noexcept { return convert_signed_integer(rhs); }
+	constexpr posit& operator=(long rhs)               noexcept { return convert_signed_integer(rhs); }
+	constexpr posit& operator=(long long rhs)          noexcept { return convert_signed_integer(rhs); }
+	constexpr posit& operator=(char rhs)               noexcept { return convert_unsigned_integer(static_cast<unsigned char>(rhs)); }
+	constexpr posit& operator=(unsigned short rhs)     noexcept { return convert_unsigned_integer(rhs); }
+	constexpr posit& operator=(unsigned int rhs)       noexcept { return convert_unsigned_integer(rhs); }
+	constexpr posit& operator=(unsigned long rhs)      noexcept { return convert_unsigned_integer(rhs); }
+	constexpr posit& operator=(unsigned long long rhs) noexcept { return convert_unsigned_integer(rhs); }
 	CONSTEXPRESSION posit& operator=(float rhs) noexcept { return convert_ieee754(rhs); }
 	CONSTEXPRESSION posit& operator=(double rhs) noexcept { return convert_ieee754(rhs); }
 
@@ -1202,6 +1201,142 @@ private:
 		return s * r * e * f;
 	}
 	
+	// Encode a positive integer magnitude into the posit bit pattern as a uint64_t,
+	// returning a saturation flag in `saturated` (true means we hit maxpos and the caller
+	// should not apply rounding twice). All work is on uint64_t, so this is constexpr-clean
+	// for nbits <= 64.
+	static constexpr uint64_t encode_positive_uint64(uint64_t v, bool& saturated) noexcept {
+		saturated = false;
+		if (v == 0ull) return 0ull;
+
+		int scale = static_cast<int>(find_msb(v)) - 1;
+
+		if (check_inward_projection_range<nbits, es, bt>(scale)) {
+			saturated = true;
+			// maxpos encoding: 0 sign + (nbits-1) ones
+			constexpr unsigned tw = nbits - 1u;
+			return (tw >= 64u) ? ~0ull : ((1ull << tw) - 1ull);
+		}
+
+		constexpr unsigned target_width = nbits - 1u; // payload width (excludes sign)
+		int k = scale >> es;
+		uint64_t exp_field = static_cast<uint64_t>(scale & static_cast<int>((1u << es) - 1u));
+
+		// Encoding layout (msb-first within target_width payload bits):
+		//   [k+1 ones] [0 terminator] [es exp bits] [scale fraction bits]
+		unsigned regime_ones = static_cast<unsigned>(k + 1);
+		unsigned ones_to_place = (regime_ones < target_width) ? regime_ones : target_width;
+
+		uint64_t encoded = ((1ull << ones_to_place) - 1ull) << (target_width - ones_to_place);
+
+		if (ones_to_place < target_width) {
+			// Terminator zero is implicitly present at position (target_width - ones_to_place - 1)
+			unsigned bits_after_regime = target_width - ones_to_place - 1u; // -1 for terminator
+			unsigned exp_bits_to_place = (es < bits_after_regime) ? static_cast<unsigned>(es) : bits_after_regime;
+			if (exp_bits_to_place > 0u) {
+				uint64_t exp_top = (es > exp_bits_to_place) ? (exp_field >> (es - exp_bits_to_place)) : exp_field;
+				encoded |= exp_top << (bits_after_regime - exp_bits_to_place);
+			}
+
+			unsigned bits_after_exp = bits_after_regime - exp_bits_to_place;
+			unsigned scale_u = static_cast<unsigned>(scale);
+			unsigned frac_to_embed = (scale_u < bits_after_exp) ? scale_u : bits_after_exp;
+			if (frac_to_embed > 0u) {
+				uint64_t frac_value = (v >> (scale_u - frac_to_embed)) & ((1ull << frac_to_embed) - 1ull);
+				// Fraction is MSB-aligned within its field; pad LSBs with zeros if scale < bits_after_exp
+				encoded |= frac_value << (bits_after_exp - frac_to_embed);
+			}
+
+			// Round-to-nearest-even on the bits we discarded.
+			// Discarded sequence (msb-first):
+			//   low (es - exp_bits_to_place) bits of exp_field, then low (scale - frac_to_embed) bits of v
+			bool round_bit = false;
+			bool sticky_bit = false;
+
+			unsigned discarded_exp = static_cast<unsigned>(es) - exp_bits_to_place;
+			unsigned discarded_frac = scale_u - frac_to_embed;
+
+			if (discarded_exp > 0u) {
+				round_bit = ((exp_field >> (discarded_exp - 1u)) & 1ull) != 0ull;
+				if (discarded_exp > 1u) {
+					uint64_t low_exp_mask = (1ull << (discarded_exp - 1u)) - 1ull;
+					if ((exp_field & low_exp_mask) != 0ull) sticky_bit = true;
+				}
+				if (scale_u > 0u) {
+					uint64_t frac_mask = (scale_u >= 64u) ? ~0ull : ((1ull << scale_u) - 1ull);
+					if ((v & frac_mask) != 0ull) sticky_bit = true;
+				}
+			}
+			else if (discarded_frac > 0u) {
+				round_bit = ((v >> (discarded_frac - 1u)) & 1ull) != 0ull;
+				if (discarded_frac > 1u) {
+					uint64_t low_frac_mask = (1ull << (discarded_frac - 1u)) - 1ull;
+					if ((v & low_frac_mask) != 0ull) sticky_bit = true;
+				}
+			}
+
+			bool last_bit = (encoded & 1ull) != 0ull;
+			if (round_bit && (sticky_bit || last_bit)) {
+				++encoded;
+				// Rounding could have overflowed past target_width into the sign bit.
+				// In that case the integer rounded up to maxpos's neighbor; saturate.
+				if ((encoded >> target_width) != 0ull) {
+					saturated = true;
+					return (target_width >= 64u) ? ~0ull : ((1ull << target_width) - 1ull);
+				}
+			}
+		}
+		return encoded;
+	}
+
+	// Constexpr integer-to-posit conversion for nbits <= 64.
+	// Builds the posit encoding directly via uint64_t shift/mask, bypassing convert_<>()
+	// (which is not constexpr because it relies on non-constexpr blockbinary operators).
+	// For nbits > 64, the existing IEEE-754 path is used at runtime.
+	template<typename Ty>
+	constexpr posit<nbits, es, bt>& convert_unsigned_integer(Ty rhs) noexcept {
+		clear();
+		if (rhs == Ty(0)) return *this;
+
+		if constexpr (nbits > 64) {
+			return convert_ieee754(static_cast<double>(rhs));
+		}
+		else {
+			bool saturated = false;
+			uint64_t encoded = encode_positive_uint64(static_cast<uint64_t>(rhs), saturated);
+			setbits(encoded);
+			return *this;
+		}
+	}
+
+	template<typename Ty>
+	constexpr posit<nbits, es, bt>& convert_signed_integer(Ty rhs) noexcept {
+		clear();
+		if (rhs == Ty(0)) return *this;
+
+		if constexpr (nbits > 64) {
+			return convert_ieee754(static_cast<double>(rhs));
+		}
+		else {
+			bool s = (rhs < Ty(0));
+			using UnsignedTy = std::make_unsigned_t<Ty>;
+			UnsignedTy abs_v = s
+				? static_cast<UnsignedTy>(UnsignedTy(0) - static_cast<UnsignedTy>(rhs))
+				: static_cast<UnsignedTy>(rhs);
+
+			bool saturated = false;
+			uint64_t encoded = encode_positive_uint64(static_cast<uint64_t>(abs_v), saturated);
+			if (s) {
+				// Posit negation = two's complement on the full nbits encoding.
+				// Done in uint64_t to avoid the non-constexpr blockbinary::operator+= chain.
+				constexpr uint64_t nbits_mask = (nbits >= 64u) ? ~uint64_t(0) : ((uint64_t(1) << nbits) - 1ull);
+				encoded = (~encoded + 1ull) & nbits_mask;
+			}
+			setbits(encoded);
+			return *this;
+		}
+	}
+
 	template <typename Real>
 	CONSTEXPRESSION posit<nbits, es, bt>& convert_ieee754(const Real& rhs) noexcept {
 		// Direct IEEE-754 to posit conversion using frexp to extract scale and fraction.

--- a/include/sw/universal/number/posit/posit_impl.hpp
+++ b/include/sw/universal/number/posit/posit_impl.hpp
@@ -582,7 +582,17 @@ public:
 	constexpr posit& operator=(int rhs)                noexcept { return convert_signed_integer(rhs); }
 	constexpr posit& operator=(long rhs)               noexcept { return convert_signed_integer(rhs); }
 	constexpr posit& operator=(long long rhs)          noexcept { return convert_signed_integer(rhs); }
-	constexpr posit& operator=(char rhs)               noexcept { return convert_unsigned_integer(static_cast<unsigned char>(rhs)); }
+	constexpr posit& operator=(char rhs)               noexcept {
+		// Plain char is implementation-defined as either signed or unsigned;
+		// dispatch to the matching conversion so negative values on signed-char
+		// platforms (the common case) sign-extend correctly.
+		if constexpr (std::is_signed_v<char>) {
+			return convert_signed_integer(rhs);
+		}
+		else {
+			return convert_unsigned_integer(static_cast<unsigned char>(rhs));
+		}
+	}
 	constexpr posit& operator=(unsigned short rhs)     noexcept { return convert_unsigned_integer(rhs); }
 	constexpr posit& operator=(unsigned int rhs)       noexcept { return convert_unsigned_integer(rhs); }
 	constexpr posit& operator=(unsigned long rhs)      noexcept { return convert_unsigned_integer(rhs); }

--- a/static/tapered/posit/api/api.cpp
+++ b/static/tapered/posit/api/api.cpp
@@ -40,6 +40,56 @@ try {
 	/////////////////////////////////////////////////////////////////////////////////////
 	//// posit construction, initialization, assignment and comparisions
 
+	std::cout << "+-----------------   constexpr integer construction (issue #713)\n";
+	{
+		// constexpr construction from integer literals must succeed at compile time
+		// for nbits <= 64. The encoded bit pattern must match the runtime path exactly.
+		constexpr posit<32, 2>  cx_pos_42(42);
+		constexpr posit<32, 2>  cx_neg_42(-42);
+		constexpr posit<32, 2>  cx_zero(0);
+		constexpr posit<8,  0>  cx_three(3);
+		constexpr posit<16, 1>  cx_kilo(1024);
+		constexpr posit<64, 3>  cx_big(123456789LL);
+		constexpr posit<8,  0>  cx_sat_pos(1000);   // saturates to maxpos
+		constexpr posit<8,  0>  cx_sat_neg(-1000);  // saturates to maxneg
+		constexpr posit<32, 2>  cx_int_min(int32_t(-2147483647 - 1));
+
+		// runtime construction for cross-check
+		posit<32, 2>  rt_pos_42; rt_pos_42 = 42;
+		posit<32, 2>  rt_neg_42; rt_neg_42 = -42;
+		posit<32, 2>  rt_zero;   rt_zero   = 0;
+		posit<8,  0>  rt_three;  rt_three  = 3;
+		posit<16, 1>  rt_kilo;   rt_kilo   = 1024;
+		posit<64, 3>  rt_big;    rt_big    = 123456789LL;
+		posit<8,  0>  rt_sat_pos; rt_sat_pos = 1000;
+		posit<8,  0>  rt_sat_neg; rt_sat_neg = -1000;
+		posit<32, 2>  rt_int_min; rt_int_min = int32_t(-2147483647 - 1);
+
+		auto same_bits = [](auto cx, auto rt) {
+			auto a = cx.bits();
+			auto b = rt.bits();
+			constexpr unsigned nrBlocks = decltype(a)::nrBlocks;
+			for (unsigned i = 0; i < nrBlocks; ++i) {
+				if (a.block(i) != b.block(i)) return false;
+			}
+			return true;
+		};
+
+		int start = nrOfFailedTestCases;
+		if (!same_bits(cx_pos_42,  rt_pos_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(42)\n"; }
+		if (!same_bits(cx_neg_42,  rt_neg_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(-42)\n"; }
+		if (!same_bits(cx_zero,    rt_zero))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(0)\n"; }
+		if (!same_bits(cx_three,   rt_three))   { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(3)\n"; }
+		if (!same_bits(cx_kilo,    rt_kilo))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<16,1>(1024)\n"; }
+		if (!same_bits(cx_big,     rt_big))     { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<64,3>(123456789)\n"; }
+		if (!same_bits(cx_sat_pos, rt_sat_pos)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(1000) sat\n"; }
+		if (!same_bits(cx_sat_neg, rt_sat_neg)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(-1000) sat\n"; }
+		if (!same_bits(cx_int_min, rt_int_min)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(INT_MIN)\n"; }
+		if (nrOfFailedTestCases - start == 0) {
+			std::cout << "PASS constexpr integer construction\n";
+		}
+	}
+
 	std::cout << "+-----------------   posit construction, initialization, comparisons\n";
 	{
 		int start = nrOfFailedTestCases;

--- a/static/tapered/posit/api/api.cpp
+++ b/static/tapered/posit/api/api.cpp
@@ -43,7 +43,11 @@ try {
 	std::cout << "+-----------------   constexpr integer construction (issue #713)\n";
 	{
 		// constexpr construction from integer literals must succeed at compile time
-		// for nbits <= 64. The encoded bit pattern must match the runtime path exactly.
+		// for nbits <= 64. The encoded bit pattern must match an INDEPENDENT
+		// reference path (convert_ieee754 via double-cast) so that a bug in
+		// encode_positive_uint64 cannot mask itself.
+		// All test values fit exactly in double's 53-bit mantissa, so the double
+		// cast is bit-exact and provides a valid reference.
 		constexpr posit<32, 2>  cx_pos_42(42);
 		constexpr posit<32, 2>  cx_neg_42(-42);
 		constexpr posit<32, 2>  cx_zero(0);
@@ -54,20 +58,21 @@ try {
 		constexpr posit<8,  0>  cx_sat_neg(-1000);  // saturates to maxneg
 		constexpr posit<32, 2>  cx_int_min(int32_t(-2147483647 - 1));
 
-		// runtime construction for cross-check
-		posit<32, 2>  rt_pos_42; rt_pos_42 = 42;
-		posit<32, 2>  rt_neg_42; rt_neg_42 = -42;
-		posit<32, 2>  rt_zero;   rt_zero   = 0;
-		posit<8,  0>  rt_three;  rt_three  = 3;
-		posit<16, 1>  rt_kilo;   rt_kilo   = 1024;
-		posit<64, 3>  rt_big;    rt_big    = 123456789LL;
-		posit<8,  0>  rt_sat_pos; rt_sat_pos = 1000;
-		posit<8,  0>  rt_sat_neg; rt_sat_neg = -1000;
-		posit<32, 2>  rt_int_min; rt_int_min = int32_t(-2147483647 - 1);
+		// Reference path: float literal constructor still routes through the
+		// pre-existing convert_ieee754, independent of the new constexpr code.
+		posit<32, 2>  ref_pos_42 (42.0);
+		posit<32, 2>  ref_neg_42 (-42.0);
+		posit<32, 2>  ref_zero   (0.0);
+		posit<8,  0>  ref_three  (3.0);
+		posit<16, 1>  ref_kilo   (1024.0);
+		posit<64, 3>  ref_big    (123456789.0);
+		posit<8,  0>  ref_sat_pos(1000.0);
+		posit<8,  0>  ref_sat_neg(-1000.0);
+		posit<32, 2>  ref_int_min(static_cast<double>(int32_t(-2147483647 - 1)));
 
-		auto same_bits = [](auto cx, auto rt) {
+		auto same_bits = [](auto cx, auto ref) {
 			auto a = cx.bits();
-			auto b = rt.bits();
+			auto b = ref.bits();
 			constexpr unsigned nrBlocks = decltype(a)::nrBlocks;
 			for (unsigned i = 0; i < nrBlocks; ++i) {
 				if (a.block(i) != b.block(i)) return false;
@@ -76,15 +81,25 @@ try {
 		};
 
 		int start = nrOfFailedTestCases;
-		if (!same_bits(cx_pos_42,  rt_pos_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(42)\n"; }
-		if (!same_bits(cx_neg_42,  rt_neg_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(-42)\n"; }
-		if (!same_bits(cx_zero,    rt_zero))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(0)\n"; }
-		if (!same_bits(cx_three,   rt_three))   { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(3)\n"; }
-		if (!same_bits(cx_kilo,    rt_kilo))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<16,1>(1024)\n"; }
-		if (!same_bits(cx_big,     rt_big))     { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<64,3>(123456789)\n"; }
-		if (!same_bits(cx_sat_pos, rt_sat_pos)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(1000) sat\n"; }
-		if (!same_bits(cx_sat_neg, rt_sat_neg)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(-1000) sat\n"; }
-		if (!same_bits(cx_int_min, rt_int_min)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(INT_MIN)\n"; }
+		if (!same_bits(cx_pos_42,  ref_pos_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(42)\n"; }
+		if (!same_bits(cx_neg_42,  ref_neg_42))  { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(-42)\n"; }
+		if (!same_bits(cx_zero,    ref_zero))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(0)\n"; }
+		if (!same_bits(cx_three,   ref_three))   { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(3)\n"; }
+		if (!same_bits(cx_kilo,    ref_kilo))    { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<16,1>(1024)\n"; }
+		if (!same_bits(cx_big,     ref_big))     { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<64,3>(123456789)\n"; }
+		if (!same_bits(cx_sat_pos, ref_sat_pos)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(1000) sat\n"; }
+		if (!same_bits(cx_sat_neg, ref_sat_neg)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(-1000) sat\n"; }
+		if (!same_bits(cx_int_min, ref_int_min)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(INT_MIN)\n"; }
+
+		// Plain-char regression: on signed-char platforms (the common case),
+		// posit<8,0>(char(-3)) must encode -3, not 253.
+		constexpr posit<32, 2> cx_neg_char(static_cast<char>(-3));
+		posit<32, 2> ref_neg_char(-3.0);
+		if (!same_bits(cx_neg_char, ref_neg_char)) {
+			++nrOfFailedTestCases;
+			std::cout << "FAIL constexpr posit<32,2>(char(-3)) - char signedness dispatch\n";
+		}
+
 		if (nrOfFailedTestCases - start == 0) {
 			std::cout << "PASS constexpr integer construction\n";
 		}

--- a/static/tapered/posit/api/api.cpp
+++ b/static/tapered/posit/api/api.cpp
@@ -91,13 +91,25 @@ try {
 		if (!same_bits(cx_sat_neg, ref_sat_neg)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<8,0>(-1000) sat\n"; }
 		if (!same_bits(cx_int_min, ref_int_min)) { ++nrOfFailedTestCases; std::cout << "FAIL constexpr posit<32,2>(INT_MIN)\n"; }
 
-		// Plain-char regression: on signed-char platforms (the common case),
-		// posit<8,0>(char(-3)) must encode -3, not 253.
-		constexpr posit<32, 2> cx_neg_char(static_cast<char>(-3));
-		posit<32, 2> ref_neg_char(-3.0);
-		if (!same_bits(cx_neg_char, ref_neg_char)) {
-			++nrOfFailedTestCases;
-			std::cout << "FAIL constexpr posit<32,2>(char(-3)) - char signedness dispatch\n";
+		// Plain-char regression: char is implementation-defined as signed or
+		// unsigned, so the reference value must match the platform's char model.
+		// On signed-char platforms (the common case) char(-3) is -3; on
+		// unsigned-char platforms it wraps to 253. The dispatch in operator=(char)
+		// is what we are validating: it should produce the matching reference.
+		constexpr posit<32, 2> cx_char_neg3(static_cast<char>(-3));
+		if constexpr (std::is_signed_v<char>) {
+			posit<32, 2> ref_char_neg3(-3.0);
+			if (!same_bits(cx_char_neg3, ref_char_neg3)) {
+				++nrOfFailedTestCases;
+				std::cout << "FAIL constexpr posit<32,2>(char(-3)) on signed-char platform\n";
+			}
+		}
+		else {
+			posit<32, 2> ref_char_neg3(253.0);
+			if (!same_bits(cx_char_neg3, ref_char_neg3)) {
+				++nrOfFailedTestCases;
+				std::cout << "FAIL constexpr posit<32,2>(char(-3)) on unsigned-char platform\n";
+			}
 		}
 
 		if (nrOfFailedTestCases - start == 0) {

